### ISSUE TITLE
Review ElementCacheMap

### ISSUE THREAD

**Issue by jbrezmorf** [flow123d/flow123d#1205]
- review ElementCacheMap after we make new assembly scheme working
- plenty of space for simplification and improvements there
- some TODOs added in  d540fc2

**Comment by jbrezmorf** [comment]
Further tips for improvement (after #1259).

Basic algorithm for creation of ElementCacheMap:
1. collect evaluation points of the elements of the patch (i_reg, i_element, i_eval_point)
2. sort these by region and element
3. determine begins of regions in the value cache and in the element cache map
4. set `element_cache_map[i_element][i_eval_point] = position in the sorted list` 

Currently the sort is performed by the bucket sort using the map from region ids to the lists of elements.
This is possibly not better then actually sort the array by a standard sorting algorithm.

Tips:
- try to use simpler sorting mechanism
- move creation of the element cache map possibly into a method of generic assembly making the temporary structures 
    a local data on the stack 
- use sorted list `(i_reg, i_element, i_eval_point)` as the inverse element cache map, can be used to clear the element cache map
- postpone these optimization after we simplify PointBase and its descendants
- see if we can use sort of compressed element cache map, that is do not have dense elements x points table, but its sparse representation; that should be possible

**Comment by jbrezmorf** [comment]
Postpone after application to all equations and test on real benchmarks.

**Comment by jbrezmorf** [comment]
Done.
